plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergUtil.java

```diff
@@ -13,36 +13,170 @@
 import static io.trino.plugin.hive.HiveMetadata.PRESTO_VIEW_EXPANDED_TEXT_MARKER;
+import static io.trino.plugin.hive.HiveMetadata.TABLE_COMMENT;
 import static io.trino.plugin.hive.TableType.EXTERNAL_TABLE;
 import static io.trino.plugin.hive.TableType.VIRTUAL_VIEW;
 import static io.trino.plugin.hive.ViewReaderUtil.ICEBERG_MATERIALIZED_VIEW_COMMENT;
+import static io.trino.plugin.iceberg.catalog.AbstractTrinoCatalog.COLUMN_TRINO_TYPE_PROPERTY;
+import static io.trino.plugin.iceberg.catalog.AbstractTrinoCatalog.TRINO_TABLE_METADATA_INFO_VALID_FOR;
 import static java.util.Locale.ENGLISH;
 import static org.apache.iceberg.BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE;
+import static org.apache.iceberg.BaseMetastoreTableOperations.METADATA_LOCATION_PROP;
 import static org.apache.iceberg.BaseMetastoreTableOperations.TABLE_TYPE_PROP;
 
 public final class GlueIcebergUtil
 {
     private GlueIcebergUtil() {}
 
-    public static TableInput getTableInput(String tableName, Optional<String> owner, Map<String, String> parameters)
+    private static final int GLUE_TABLE_PARAMETER_LENGTH_LIMIT = 512000;
+    private static final int GLUE_COLUMN_TYPE_LENGTH_LIMIT = 131072;
+    private static final int GLUE_COLUMN_PARAMETER_LENGTH_LIMIT = 512000;
+
+    public static TableInput getTableInput(
+            TypeManager typeManager,
+            String tableName,
+            Optional<String> owner,
+            TableMetadata metadata,
+            String newMetadataLocation,
+            Map<String, String> parameters,
+            boolean cacheTableMetadata)
     {
-        return new TableInput()
+        parameters = new HashMap<>(parameters);
+        parameters.putIfAbsent(TABLE_TYPE_PROP, ICEBERG_TABLE_TYPE_VALUE.toUpperCase(ENGLISH));
+        parameters.put(METADATA_LOCATION_PROP, newMetadataLocation);
+        parameters.remove(TRINO_TABLE_METADATA_INFO_VALID_FOR); // no longer valid
+
+        TableInput tableInput = new TableInput()
                 .withName(tableName)
                 .withOwner(owner.orElse(null))
-                .withParameters(ImmutableMap.<String, String>builder()
-                        .putAll(parameters)
-                        .put(TABLE_TYPE_PROP, ICEBERG_TABLE_TYPE_VALUE.toUpperCase(ENGLISH))
-                        .buildKeepingLast())
                 // Iceberg does not distinguish managed and external tables, all tables are treated the same and marked as EXTERNAL
                 .withTableType(EXTERNAL_TABLE.name());
+
+        if (cacheTableMetadata) {
+            // Store table metadata sufficient to answer information_schema.columns and system.metadata.table_comments queries, which are often queried in bulk by e.g. BI tools
+            String comment = metadata.properties().get(TABLE_COMMENT);
+            Optional<List<Column>> glueColumns = glueColumns(typeManager, metadata);
+
+            boolean canPersistComment = (comment == null || comment.length() <= GLUE_TABLE_PARAMETER_LENGTH_LIMIT);
```

Comment:
It would be better to check column name length as well. Previously, we could create tables with columns having > 255 characters, but the query fails after this change.